### ISSUE TITLE
How to validate subschema (like a schema in openapi)? 

### ISSUE THREAD

**Issue by AtosNicoS** [gregsdennis/Manatee.Json#246]
Hi!
I've written an OpenApi Spec for my Rest API. I want to validate the payload inside the server and the client test suit. The idea was to validate against the existing schemas inside the openapi spec.

But how do I reference a subschema? In My example I would use `openapi.json/components/schemas/Test`, but the referenced `TestId` is inside the same file. So I need to read the *whole file* to *parse the schema provided as json path*. How can I do this? If this is not possible with this library, does anyone know an alternative (preferred in c#).

```json
[...]
"components": {
    "schemas": {
        "Test": {
            "description": "Test Object",
            "type": "object",
            "additionalProperties": false,
            "properties": {
                "firmwareVersion": {
                    "type": "string",
                    "description": "Version of the Firmware",
                    "example": "1.1.4"
                },
                "id": {
                    "$ref": "#/components/schemas/TestId"
                }
            }
        },
        "TestId": {
            "type": "string",
            "format": "uuid",
            "description": "TestId"
        }
    }
}
```

**Comment by gregsdennis** [comment]
I think you're talking about [this](https://github.com/OAI/OpenAPI-Specification/blob/master/schemas/v2.0/schema.json)?  That *is* a fairly large schema.  

Yes, you're going to have to parse the entire file before being able to use it.  However, once the model is in memory, you can use it.

```c#
var text = // content of the file
var json = JsonValue.Parse(text);
var serializer = new JsonSerializer();
var schema = serializer.Deserialize<JsonSchema>(json);
```

If you hold on to `schema` in a static field, it'll persist between calls to your Rest API.

There is also a version of `Parse()` that takes a stream.

Does that answer your question?

**Comment by AtosNicoS** [comment]
I think that is not a solution for me.

1. I need to validate against an openapi subschema. I do not want to validate, if my file is a valid openapi schema (the one you linked), I do have a valid openapi schema, that contains other schemas I want to validate. See [this example](https://github.com/OAI/OpenAPI-Specification/blob/master/examples/v3.0/petstore-expanded.yaml#L127). In this case I want to validate if my json has a property name (from the ref) and a property id.
2. Openapi schema are parsed a bit different to json schemas. They have a n"nullable" property for example, so that must also be respected.

**Comment by gregsdennis** [comment]
I think it might be beneficial to come up with a smaller example of what you're trying to do.

It sounds like you have JSON that you want to validate against a subschema of the Open API schema (that I linked). In order to obtain that subschema you need to first read the root schema.

There is a lot of current work to get Open API 3 to confirm to JSON Schema draft 2019-09 by defining a new vocabulary to define those other keywords.  Until then, you can define them yourself and add them to Manatee.Json.

**Comment by AtosNicoS** [comment]
Imagine this schema:
```json
{
  "components": {
    "schemas": {
      "Pet": {
        "allOf": [
          {
            "$ref": "#/components/schemas/NewPet"
          },
          {
            "type": "object",
            "required": [
              "id"
            ],
            "properties": {
              "id": {
                "type": "integer",
                "format": "int64"
              }
            }
          }
        ]
      },
      "NewPet": {
        "type": "object",
        "required": [
          "name"
        ],
        "properties": {
          "name": {
            "type": "string"
          },
          "tag": {
            "type": "string"
          }
        }
      }
    }
  }
}
```

And this data:
```json
{
"name": "TestPet",
"tag": "cat",
"id": 32
}
```

And this invalid data
```json
{
"nickname": "TestPet2",
"tag": "cat",
"id": 64
}
```

Ecept there are some openapi specific properties that will get resolved in 3.1 (as you explained), how would I validate this json with the provided schema? I'd need to validate against the schema path `components/schemas/Pet`.

**Comment by gregsdennis** [comment]
Okay.  I think I see your problem. Unfortunately, I don't think it's one that any library is going to be able to solve without some work on your end.

The main issue is that `components` is not a recognized JSON Schema keyword (in any draft), so the spec says to skip it.

Additionally, if you isolate the `Pet` subschema and validate on just that, the subschema becomes the root, and it won't be able to resolve the `#/components/schemas/NewPet` reference because that requires the resolution logic to navigate outside of the root schema.  I think this problem is one that @handrews was working so hard to resolve by creating the `$recursiveRef` and `$recursiveAnchor` keywords for draft 2019-09.

The only solution I can think of is storing the `Pet` and `NewPet` schemas in their own files and giving them unique `$id`s.  Then you can reference them directly.  (You'll need to register them with `JsonSchemaRegistry` before you attempt to validate. See [the docs](https://gregsdennis.github.io/Manatee.Json/usage/schema/references.html#schema-registration) for more info on registration.)
